ts/score/peerScore.ts

```diff
@@ -0,0 +1,596 @@
+import { Message } from '../message'
+import { PeerScoreParams, validatePeerScoreParams } from './peerScoreParams'
+import { PeerStats, createPeerStats, ensureTopicStats } from './peerStats'
+import { computeScore } from './computeScore'
+import { MessageDeliveries, DeliveryRecordStatus } from './messageDeliveries'
+import PeerId = require('peer-id')
+import Multiaddr = require('multiaddr')
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+// @ts-ignore
+import debug = require('debug')
+
+const log = debug('libp2p:gossipsub:score')
+
+interface AddressBook {
+  getMultiaddrsForPeer(id: PeerId): Multiaddr[]
+  // eslint-disable-next-line @typescript-eslint/ban-types
+  on(evt: string, fn: Function): void
+  // eslint-disable-next-line @typescript-eslint/ban-types
+  off(evt: string, fn: Function): void
+}
+
+export class PeerScore {
+  /**
+   * The score parameters
+   */
+  params: PeerScoreParams
+  /**
+   * Per-peer stats for score calculation
+   */
+  peerStats: Map<PeerId, PeerStats>
```

Comment:
Right, this is currently using `PeerId` throughout, and relying on the assumption that `PeerId` objects are not copied or multiple `PeerId` objects identify the same peer.

The hope was that we could avoid an additional layer of `PeerId` => `string` conversion here, if its not needed. It kills me to see the `id.toB58String()` constantly recomputed. Though now, looking at the PeerId implementation, it seems the string value is cached, so the worry is unfounded.

Should we use string PeerId identifiers throughout this module, making all methods take a string arg instead of a PeerId? Scanning the file, it seems the only use of PeerId, as an object, is in consuming the AddressBook API. Using ConnectionManager instead, we may get the connection info we need by using `connectionManager.connections.get(peerIdStr)` directly.

Reply:
I understand the pain on the `id.toB58String()` all around as I also did that in the PeerStore. I think it is the best call and also reduces a lot of data in the keys of the data structures.
In theory, I think we will not need a PeerId instance, but if we need we can easily create one from `PeerId.createFromCID`.
So yes, I think we should just use string arg in all the methods here